pkg/river/parser/interp_string.go

```diff
@@ -0,0 +1,282 @@
+	// Start of current token.
+	pos = s.file.Pos(s.baseOffset + s.offset)
+
+	switch ch := s.ch; {
+	case ch == eof:
+		tok = interpStringTokenEOF
+		return
+
+	case ch == '$' && s.peek() == '{':
+		s.next() // Consume $.
+		s.next() // Consume {.
+
+		if text, terminated := s.scanExpr(); terminated {
+			tok = interpStringTokenExpr
+			lit = text
+		} else {
+			tok = interpStringTokenRaw
+			lit = text
+		}
+
+	default:
+		tok = interpStringTokenRaw
+		lit = s.scanRaw()
+	}
+
+	return
+}
+
+func (s *interpStringScanner) scanExpr() (fragment string, terminated bool) {
+	// s.offset is first byte after ${; subtract 2 to get the start of $.
+	off := s.offset - 2
+
+	// curlyPairs tracks the number of pairs of {} we expect. We start at 1 to
+	// count the first { from ${. We only stop scanning the expr once we've
+	// consumed the last } from all the expected pairs.
+	curlyPairs := 1
+
+	for {
+		ch := s.ch
+		if ch == eof {
+			break
+		}
+
+		s.next()
+
+		if ch == '{' {
+			curlyPairs++
+		} else if ch == '}' {
+			curlyPairs--
+			if curlyPairs == 0 {
+				terminated = true
+				break
+			}
+		}
+	}
+
+	return string(s.input[off:s.offset]), terminated
+}
+
+func (s *interpStringScanner) scanRaw() string {
+	off := s.offset
+
+	for {
+		ch := s.ch
+		if ch == eof {
+			break
+		}
+
+		if ch == '$' && s.peek() == '{' {
+			break
+		}
+
+		s.next()
+		if ch == '\\' {
+			s.scanEscape()
+		}
+	}
+
+	return string(s.input[off:s.offset])
+}
+
+// scanEscape parses an escape sequence. In case of a syntax error, scanEscape
+// stops at the offending character without consuming it.
+func (s *interpStringScanner) scanEscape() {
+	off := s.offset
+
+	var (
+		n         int
+		base, max uint32
+	)
+
+	switch s.ch {
+	case 'a', 'b', 'f', 'n', 'r', 't', 'v', '\\', '"', '$':
+		s.next()
+		return
+	case '0', '1', '2', '3', '4', '5', '6', '7':
+		n, base, max = 3, 8, 255
+	case 'x':
+		s.next()
+		n, base, max = 2, 16, 255
+	case 'u':
+		s.next()
+		n, base, max = 4, 16, unicode.MaxRune
+	case 'U':
+		s.next()
+		n, base, max = 8, 16, unicode.MaxRune
+	default:
+		msg := "unknown escape sequence"
+		if s.ch == eof {
+			msg = "escape sequence not terminated"
+		}
+		s.onError(off, msg)
+		return
+	}
+
+	var x uint32
+	for n > 0 {
+		d := uint32(digitVal(s.ch))
+		if d >= base {
+			msg := fmt.Sprintf("illegal character %#U in escape sequence", s.ch)
+			if s.ch == eof {
+				msg = "escape sequence not terminated"
+			}
+			s.onError(off, msg)
+			return
+		}
+		x = x*base + d
+		s.next()
+		n--
+	}
+
+	if x > max || x >= 0xD800 && x < 0xE000 {
+		s.onError(off, "escape sequence is invalid Unicode code point")
+	}
+}
+
+func digitVal(ch rune) int {
+	switch {
+	case ch >= '0' && ch <= '9':
+		return int(ch - '0')
+	case lower(ch) >= 'a' && lower(ch) <= 'f':
+		return int(lower(ch) - 'a' + 10)
+	}
+	return 16 // Larger than any legal digit val
+}
+
+func lower(ch rune) rune { return ('a' - 'A') | ch }
```

Comment:
These utility methods are copied verbatim from river/scanner, but I don't think it makes sense to try to refactor them into a shared location.